### ISSUE TITLE
计算机视觉实战：从图像处理到深度学习

### ISSUE THREAD

**Issue by sbumsid24** [sbumsid24/WB#4]
引言
计算机视觉是人工智能领域最活跃的分支之一，广泛应用于安防监控、自动驾驶、医疗影像等领域。本文将系统介绍计算机视觉的核心技术和实现方法，涵盖OpenCV基础操作、特征提取、目标检测和深度学习模型应用，并提供大量可运行的Python代码示例。

1. OpenCV基础图像处理
1.1 图像读写与基本操作
import cv2
import numpy as np
import matplotlib.pyplot as plt

# 读取图像
img = cv2.imread('example.jpg')
print(f"图像尺寸: {img.shape}")  # (高度, 宽度, 通道数)

# 转换为灰度图
gray = cv2.cvtColor(img, cv2.COLOR_BGR2GRAY)

# 显示图像
plt.figure(figsize=(10, 5))
plt.subplot(121), plt.imshow(cv2.cvtColor(img, cv2.COLOR_BGR2RGB)), plt.title('原始图像')
plt.subplot(122), plt.imshow(gray, cmap='gray'), plt.title('灰度图像')
plt.show()

# 保存图像
cv2.imwrite('gray_example.jpg', gray)
1.2 图像滤波与边缘检测
# 高斯模糊
blurred = cv2.GaussianBlur(gray, (5, 5), 0)

# Canny边缘检测
edges = cv2.Canny(blurred, 50, 150)

# 显示结果
plt.figure(figsize=(15, 5))
plt.subplot(131), plt.imshow(gray, cmap='gray'), plt.title('灰度图像')
plt.subplot(132), plt.imshow(blurred, cmap='gray'), plt.title('高斯模糊')
plt.subplot(133), plt.imshow(edges, cmap='gray'), plt.title('边缘检测')
plt.show()
2. 特征提取与匹配
2.1 SIFT特征提取
# 初始化SIFT检测器
sift = cv2.SIFT_create()

# 检测关键点和描述符
keypoints, descriptors = sift.detectAndCompute(gray, None)

# 绘制关键点
img_keypoints = cv2.drawKeypoints(gray, keypoints, img, flags=cv2.DRAW_MATCHES_FLAGS_DRAW_RICH_KEYPOINTS)

plt.figure(figsize=(10, 5))
plt.imshow(cv2.cvtColor(img_keypoints, cv2.COLOR_BGR2RGB))
plt.title('SIFT关键点检测')
plt.show()
2.2 特征匹配
# 读取另一张图像进行匹配
img2 = cv2.imread('example2.jpg')
gray2 = cv2.cvtColor(img2, cv2.COLOR_BGR2GRAY)
keypoints2, descriptors2 = sift.detectAndCompute(gray2, None)

# 使用FLANN匹配器
FLANN_INDEX_KDTREE = 1
index_params = dict(algorithm=FLANN_INDEX_KDTREE, trees=5)
search_params = dict(checks=50)

flann = cv2.FlannBasedMatcher(index_params, search_params)
matches = flann.knnMatch(descriptors, descriptors2, k=2)

# 筛选优质匹配
good_matches = []
for m, n in matches:
    if m.distance < 0.7 * n.distance:
        good_matches.append(m)

# 绘制匹配结果
img_matches = cv2.drawMatches(gray, keypoints, gray2, keypoints2, good_matches, None, flags=2)

plt.figure(figsize=(15, 10))
plt.imshow(img_matches)
plt.title('特征匹配结果')
plt.show()
3. 目标检测与识别
3.1 Haar级联分类器
# 加载预训练的人脸检测模型
face_cascade = cv2.CascadeClassifier(cv2.data.haarcascades + 'haarcascade_frontalface_default.xml')

# 检测人脸
faces = face_cascade.detectMultiScale(gray, scaleFactor=1.1, minNeighbors=5, minSize=(30, 30))

# 绘制检测框
for (x, y, w, h) in faces:
    cv2.rectangle(img, (x, y), (x+w, y+h), (255, 0, 0), 2)

plt.figure(figsize=(10, 5))
plt.imshow(cv2.cvtColor(img, cv2.COLOR_BGR2RGB))
plt.title('人脸检测结果')
plt.show()
3.2 YOLO目标检测
# 加载YOLO模型
net = cv2.dnn.readNet('yolov3.weights', 'yolov3.cfg')
classes = []
with open('coco.names', 'r') as f:
    classes = [line.strip() for line in f.readlines()]

# 获取输出层
layer_names = net.getLayerNames()
output_layers = [layer_names[i[0] - 1] for i in net.getUnconnectedOutLayers()]

# 预处理图像
blob = cv2.dnn.blobFromImage(img, 0.00392, (416, 416), (0, 0, 0), True, crop=False)
net.setInput(blob)
outs = net.forward(output_layers)

# 解析检测结果
class_ids = []
confidences = []
boxes = []
for out in outs:
    for detection in out:
        scores = detection[5:]
        class_id = np.argmax(scores)
        confidence = scores[class_id]
        if confidence > 0.5:
            center_x = int(detection[0] * img.shape[1])
            center_y = int(detection[1] * img.shape[0])
            w = int(detection[2] * img.shape[1])
            h = int(detection[3] * img.shape[0])
            x = int(center_x - w / 2)
            y = int(center_y - h / 2)
            boxes.append([x, y, w, h])
            confidences.append(float(confidence))
            class_ids.append(class_id)

# 非极大值抑制
indexes = cv2.dnn.NMSBoxes(boxes, confidences, 0.5, 0.4)

# 绘制检测结果
font = cv2.FONT_HERSHEY_PLAIN
for i in range(len(boxes)):
    if i in indexes:
        x, y, w, h = boxes[i]
        label = str(classes[class_ids[i]])
        color = (0, 255, 0)
        cv2.rectangle(img, (x, y), (x + w, y + h), color, 2)
        cv2.putText(img, label, (x, y + 30), font, 2, color, 2)

plt.figure(figsize=(15, 10))
plt.imshow(cv2.cvtColor(img, cv2.COLOR_BGR2RGB))
plt.title('YOLO目标检测结果')
plt.show()
4. 深度学习图像分类
4.1 CNN模型构建与训练
import tensorflow as tf
from tensorflow.keras import layers, models

# 构建CNN模型
model = models.Sequential([
    layers.Conv2D(32, (3, 3), activation='relu', input_shape=(224, 224, 3)),
    layers.MaxPooling2D((2, 2)),
    layers.Conv2D(64, (3, 3), activation='relu'),
    layers.MaxPooling2D((2, 2)),
    layers.Conv2D(128, (3, 3), activation='relu'),
    layers.MaxPooling2D((2, 2)),
    layers.Flatten(),
    layers.Dense(128, activation='relu'),
    layers.Dense(10, activation='softmax')
])

model.compile(optimizer='adam',
              loss='sparse_categorical_crossentropy',
              metrics=['accuracy'])

# 数据增强
train_datagen = tf.keras.preprocessing.image.ImageDataGenerator(
    rescale=1./255,
    rotation_range=20,
    width_shift_range=0.2,
    height_shift_range=0.2,
    horizontal_flip=True)

# 加载数据集
train_generator = train_datagen.flow_from_directory(
    'dataset/train',
    target_size=(224, 224),
    batch_size=32,
    class_mode='sparse')

# 训练模型
history = model.fit(
    train_generator,
    epochs=10,
    validation_data=val_generator)

# 评估模型
test_loss, test_acc = model.evaluate(test_generator)
print(f'测试准确率: {test_acc:.2f}')
https://fenshouwanhui.com/baike/3196.html
https://fenshouwanhui.com/baike/3197.html
https://fenshouwanhui.com/baike/3198.html
https://fenshouwanhui.com/baike/3199.html
https://fenshouwanhui.com/baike/3200.html
https://fenshouwanhui.com/baike/3201.html
https://bazizhouyi.com/baike/2574.html
https://bazizhouyi.com/baike/2625.html
https://bazizhouyi.com/baike/2626.html
https://bazizhouyi.com/baike/2627.html
https://bazizhouyi.com/baike/2628.html
https://bazizhouyi.com/baike/2629.html
https://qingganzx.com/baike/3127.html
https://qingganzx.com/baike/3128.html
https://qingganzx.com/baike/3129.html
https://qingganzx.com/baike/3130.html
https://qingganzx.com/baike/3131.html
https://qingganzx.com/baike/3132.html
